### DISCUSSION TITLE
converting distorted newspaper strip to better looking images

### DISCUSSION THREAD

**Discussion by thisisnottaken** [ImageMagick/ImageMagick#3798] [Digital Image Processing]
Hi,
I like to read newspaper comic strip and I am sharing it at http://worldforallofus.blogspot.com/2021/04/calvin-and-hobbes.html
Currently I am making snapshots presentable by clipping and transforming using https://github.com/ImageMagick/ImageMagick/discussions/2760#discussioncomment-110597

![calvin-club-no-quad](https://user-images.githubusercontent.com/19471922/121053945-82218e80-c7d9-11eb-8a6c-b6beb24bd059.jpeg)
In above image - I am not able to choose quadrilateral so that unnecessary parts above the strip can be clipped .

Can you suggest ?

**Comment by fmw42** [comment]
Here is one way. I select the extreme corners of the comic frames as the input 4 points and then use the input image dimensions for the output 4 points. If on IM 6, use convert in place of magick.

`magick comic_strip.jpeg -distort perspective "37,76 0,0 1268,113 %w,0 1265,476 %w,%h 30,495 0,%h" comic_strip_warp.jpg`

![comic_strip_warp](https://user-images.githubusercontent.com/13458218/121061658-faa43300-c778-11eb-9ad3-2b416ecb43ec.jpg)

**Reply by thisisnottaken** [reply]
i wanted calvin and hobbes by bill watterson and the straight line above it to be in output image also. 
It seems your approach is same as https://github.com/ImageMagick/ImageMagick/discussions/3798 .
I think , this image has barrel distortion so mapping quadrilateral to rectangle was not working well .

**Reply by fmw42** [reply]
Your link to #3798 comes right back here.  Is there a proper link to another issue with this type of question?

In my answer above, then just use the end points of the line above and the bottom of the frames (or the bottom line)

`magick comic_strip.jpeg -distort perspective "35,12 0,0 1273,60 %w,0 1265,476 %w,%h 30,495 0,%h" comic_strip_warp.jpg`

![comic_strip_warp](https://user-images.githubusercontent.com/13458218/121106235-78d4f980-c7ba-11eb-9805-189df56f4c8d.jpg)

But snibgo's solution below is much better.

**Reply by thisisnottaken** [reply]
sorry . here is the link i was referring -> https://github.com/ImageMagick/ImageMagick/discussions/2760#discussioncomment-110597 .

**Comment by snibgo** [comment]
Perhaps you want to distort such that the black lines become straight and horizontal. I have a Windows BAT script for that.

First, make a version with blue lines that that you want to be straight and horizontal.

![calvinBlue](https://user-images.githubusercontent.com/33812368/121102913-a50e6b80-c7f6-11eb-982f-550d4d579735.png)

Then:
```
call %PICTBAT%str2lines3 calvinBlue.png calvin.jpeg out.png

%IMG7%magick ^
  out.png ^
  -crop 1279x439+1+61 ^
  out.jpg
```
![out](https://user-images.githubusercontent.com/33812368/121102937-ae97d380-c7f6-11eb-9874-98f85c25651c.jpg)

The method is documented at [Straightening two lines](http://im.snibgo.com/str2lines.htm).

**Reply by thisisnottaken** [reply]
This looks nice .
I am also using windows .
I will give your script a try.
Thanks.

**Reply by thisisnottaken** [reply]
hi ,
I tried your script and did not get desired output .
Here is what I did - 
(1)Downloaded http://im.snibgo.com/imsnibgoBats.zip and extracted it in d:
(2)in D:\imsnibgoBats I have two png files namely peanuts-cool.png and peanuts-coolBlue.png .
these files look like 
![peanuts-coolBlue](https://user-images.githubusercontent.com/19471922/125186368-b4466600-e247-11eb-81de-c7fc451d5939.png)
![peanuts-cool](https://user-images.githubusercontent.com/19471922/125186370-b6a8c000-e247-11eb-8aca-e7f825a8b5fb.png)
(3)I added SetLocal EnableDelayedExpansion in str2lines3.bat to avoid error ''!' is not recognized as an internal or external command,'
(4)Also I edited a line in file to set TMPDIR=temp and created folder temp
(5)Now I ran 
D:\imsnibgoBats>call str2lines3 peanuts-coolBlue.png peanuts-cool.png out.png

D:\imsnibgoBats>rem Straighten two lines: three steps combined.

D:\imsnibgoBats>SetLocal EnableDelayedExpansion

D:\imsnibgoBats>rem peanuts-coolBlue.png is image with two non-intersecting aliased blue lines.

D:\imsnibgoBats>rem peanuts-cool.png image to be distorted.

D:\imsnibgoBats>rem out.png is output.

D:\imsnibgoBats>rem  proportion of effect (typically 0.0 to 1.0) [1.0]

D:\imsnibgoBats>rem  shrink or stretch or mean

D:\imsnibgoBats>rem  prefix for environment variables [s2l3]
top=100 bot=0 sub=

But out.png looks like
![out](https://user-images.githubusercontent.com/19471922/125186493-451d4180-e248-11eb-8b52-b951881a683c.png)

**Comment by snibgo** [comment]
The blue lines must divide the image into three areas: top, middle and bottom. To do this, the lines must extend from the left side to the right side, such as this:
![peanuts-coolBlue2](https://user-images.githubusercontent.com/33812368/125198309-528ef780-e259-11eb-8e4d-3de07eabc956.png)


My published script str2lines3.bat has a bug. The conversion at the end should be:
```
%IMG7%magick ^
  %INDIST% ^
  %TMPDIR%\s2l3_lt1.miff ^
  -virtual-pixel None ^
  -compose Displace ^
    -set option:compose:args 0x100%% ^
    -composite ^
  %OUTFILE%
```
Note that the input file should be %INDIST%, not %INFILE%.

The output is then:
![out](https://user-images.githubusercontent.com/33812368/125198296-486cf900-e259-11eb-9064-c0707a2d2164.png)

**Reply by thisisnottaken** [reply]
that works . here is my final output for my blog

![cool-snoopy-fit](https://user-images.githubusercontent.com/19471922/125646312-9fedbafd-4da4-46f7-917c-551e40a101bf.png)

**Comment by thisisnottaken** [comment]
To snibgo 
Your script is very useful in handling barrel distortion .
I noticed that it strictly needs blue line file to be png . Maybe in jpeg lines do not remain aliased as needed . Can that be the reason ?

**Comment by snibgo** [comment]
My script str2lines3.bat uses a control image to distort another image. I use blue lines on the control image. In fact, the script uses values in the blue channel (the other channels don't matter). Where the pixel values in the blue channel are at the maximum, this counts as a "blue line". Where the values are not at the maximum, they are not counted as part of the blue line.

To make the control image using Gimp or any editor, copy the source image, then ensure no blue channels are at the maximum (eg by darkening the entire image slightly). Then paint the blue lines. Then save the result losslessly. DON'T save it as jpeg because that may change maximum values to something less than maximum.

Blue may not be a convenient channel, eg perhaps the image has blue sky. So you can use red or green if you prefer.

Does that answer your question?
